docs/internal/DistributedArchitectureGuide.md

```diff
@@ -12,6 +12,214 @@ A guide to the general Elasticsearch components can be found [here](https://gith
+```
+
+Where `Netty4HttpServerTransport` is a single implementation of
+`AbstractHttpServerTransport` that lives in a `transport-netty4` module. And
+security module injects SSL and headers validator.
+
+## Transport
+
+`Transport` is an umbrella term for a node-to-node communication. It's a
+TCP-based custom binary protocol. Every node in a cluster is a client and server
+at the same time. Node-to-node communication never uses HTTP transport.
+
+`Netty4Transport` is the only implementation of TCP transport and initialize
+Transport client and server. X-pack/security plugin provides secure version of
+transport with TLS and authentication - `SecurityNetty4Transport`.
+
+Once node discovers cluster it will open a pool of connections to every other
+node in a cluster, and every other node will open a pool of connections
+back. That means a connection between nodes A and B is A->B pool + B->A pool. A
+node sends requests only on connections it opens (as a client).
+
+A default connection pool is around 13 connections, a pool has sub-pools of
+connections for different purposes: ping, node-state, bulks, etc. Pool structure
+is defined in `ConnectionProfile` class.
+
+ES has resilience to disconnects, frequent reconnects, but in general we assume
+network should be stable between nodes. Nevertheless, connectivity problems
+never stop coming from SDH's and Serverless. Not every cluster has enough
+redundancy to survive prolonged network interruptions.
+
+Request timeouts are discouraged. Transport request can take all the time they
+need to finish. SO_KEEPALIVE helps to detect and tear down dead connections.
+When connection is closed with error, we close entire pool of connections and
+fail outstanding requests. Then reconnect the pool again.
+
+There are no retries on the Transport layer. Application layer should decide how
+and when to retry. Some requests don't retry, and some retry with
+`RetryableAction` or `TransportMasterNodeAction`. We might have retries in the
+Transport framework in the future with [#95100](https://github.com/elastic/elasticsearch/issues/95100).
+
+Transport does not support multiplexing. Each transport message has to be
+dispatched fully before the next message can be sent. It's recommended to
+properly size/chunk messages at application layer to provide fairness of
+delivery across many senders. There is no strict guideline for transport message
+size. Use your judgement based on: how critical message is, how often it needs
+to be delivered, is it necessary to deliver as a single large piece. Almost
+always it can be chunked.
+
+`TransportMessage` has a tall family tree. There are simple node-to-node
+messages, broadcast messages, messages that has to be acknowledge by master
+node, etc. These are important for the transport infrastructure to dispatch
+message to the right place and return response back when it's truly successful.
+
+## Snapshots
+
+Another area of different networking clients is snapshotting. ES supports
```

Comment:
There are other features such as SAML/JWT metadata reloading, watcher http action, reindex and probably ML related features such as inference that also use HTTP clients. I wonder whether we should mention them here. Or do you see snapshots being the most relevant one for the team?